### ISSUE TITLE
[FEATURE] Raw/direct api, without PUG

### ISSUE THREAD

**Issue by ncoder** [RosaeNLG/rosaenlg#72]
**Is your feature request related to a problem? Please describe.**
You've got a great library here with a ton of functionality all in one package. I can't speak for everyone, but I was interested in the ability to conjugate words, adjectives and verbs. I was off-put by the pug part of the equation. I have my own text rendering engine, and having to compile each individual sentence in a separate pug file (never mind packaging it for the browser like i need to do) didn't seem very practical.

**Describe the solution you'd like**
With a bit of hacking, I was able to use the library in a bit of a "hardcore" mode. What i'd like is to show what i've done, and ideally get this a bit cleaned up, and maybe officially supported.

Just to re-iterate, my ultimate goal is to run multi-lingual text generation, on the fly, in the browser.

I have a script that runs in node to extract a subset of the resources.

```js
const fs = require('fs');
const path = require('path')
const languageCodeGenHelper = require('rosaenlg-pug-code-gen/dist/languageCodeGenHelper')

export function extractVerbs(lang: string) {
    let verbMap = {}
    // my own code here that finds the verbs my app uses.
    return Object.keys(verbMap)
}

const argv = require('yargs').argv
const outPath = argv.out || "./assets/nlg/"
const language = argv.lang ? 'es_ES'
const verbsList = extractVerbs('es')
const wordsList = [] // similarly here,  i would populate my own word and adjectives list using my own app's data sources.
const adjectivesList = []

var codeGenHelper = languageCodeGenHelper.buildLanguageCodeGen(languageCodeGenHelper.getIso2fromLocale(language));

const verbs = codeGenHelper.getVerbsInfo(verbsList)
const words = codeGenHelper.getWordsInfo(wordsList)
const adjectives = codeGenHelper.getAdjectivesInfo(adjectivesList)


const embeddedLinguisticResources = {
    verbs,
    words,
    adjectives
};

let outFile = path.join(outPath, `resources_${languageCodeGenHelper.getIso2fromLocale(language)}.json`)
fs.writeFileSync(outFile, JSON.stringify(embeddedLinguisticResources, null, 4))
console.log("Written to " + outFile)
```

then, in the browser, you can use as:
```js

const embeddedLinguisticResources = {... (load from generated file above )... }

// define a few helpers (this is lifted out of the generated packages, would be available in lib already in some way)
...
function getAnonymous(gender, number) {
    return util.genderNumberManager.getAnonymous(gender, number);
}
function getAnonMS() {
    return util.genderNumberManager.getAnonMS();
}
function getAnonMP() {
    return util.genderNumberManager.getAnonMP();
}
function getAnonFS() {
    return util.genderNumberManager.getAnonFS();
}
function getAnonFP() {
    return util.genderNumberManager.getAnonFP();
}
....


let pug_html= ""
let spy = {
    getPugHtml: function () {
        return pug_html;
    },
    getPugMixins: function () {
        return pug_mixins;
    },
    setPugHtml: function (new_pug_html) {
        pug_html = new_pug_html;
    },
    appendPugHtml: function (append) {
        pug_html = pug_html + append;
    },
    appendDoubleSpace: function () {
        pug_html = pug_html + "  ";
    },
    getEmbeddedLinguisticResources: function () {
        return embeddedLinguisticResources;
    },

    // we should avoid this one as util. is already available
    isEvaluatingEmpty: function () {
        return util.saveRollbackManager.isEvaluatingEmpty;
    },

    isEvaluatingChoosebest: function () {
        return util.saveRollbackManager.isEvaluatingChoosebest;
    }
}

const util = new NlgLib({language})
util.setSpy(spy)

// once the boilerplate is out of the way, API usage isn't so bad like so:
// except for the fact that "r" doesn't actually contain the result.
const r = util.sentenceManager.verb(getAnonMP(), { verb: 'hablar', tense: 'INDICATIVE_IMPERFECT' }, params)

console.log(util.filterAll("el " + pug_html))
```

Code similar to this looks like it is running for me;
It looks like a few changes +  tests + documentation would enable another usecase.

Is there interest in this?

**Comment by ludans** [comment]
Thanks for the ticket and the idea.

RosaeNLG is made of 70+ packages: https://github.com/RosaeNLG/rosaenlg/tree/master/packages, which can be used independently.

Each package brings simple features (e.g. `english-a-an-list` is a list of words that take "an" vs "a" determiner, `english-a-an` is a helper on this list, etc.), while some packages like `rosaenlg-pug-*` bundle everything for Pug templates usage. `rosaenlg` run the compiled templates and makes the link with all the linguistic resources, but has no link with the Pug syntax itself.

Each package is meant to be reused independently of RosaeNLG as a whole - anybody who don't like Pug syntax but want to write a natural language generator can reuse the linguistic resources to speed up his/her development.

If you just want to conjugate verbs or agree words, you may use directly the simple packages provided, e.g. `english-plurals` and `english-plurals-list` for English words.

An issue when rendering in working in browser is that most linguistic resources are just too big. You must only bundle the verbs words adjectives you really want to render and send them to the browser, and not the whole resource. This is the job made by `rosaenlg-pug-code-gen`: it parses the Pug templates and takes only the required resources.


What I understand is that:
- you know the verbs or words to render in advance
- the multilingual part is important: using the wrapper functions of `rosaenlg` package (like `sentenceManager.verb`) is handy
- **you want a programmatic API (no templates)**

I would say that your approach is the good one (I can't think of a better one).

**I am definitely interested in moving forward. Let's think of a proper packaging.**
What about a `rosaenlg-api` package that would do the proper packaging?

**Comment by dddominikk** [comment]
Always great to see movement in this repo, this is still the most practically impressive NLG template in open source.

**Comment by ncoder** [comment]
Ludan. You are correct about your understanding of my request (referring to the list of 3 things you outlined).

I've gained a bit more information since the last message. I'd like to add more color to the discussion. You are right that rosangl is a collection of packages. What is nice is that they have a very similar API, which is good.  

I was able to make a simple unified API for verb conjugations with just spanish and english with just a few lines.
a little bit like this:

en/index.ts:
```ts
const EnglishVerbs = require('english-verbs-helper');
const Irregular = require('english-verbs-irregular/dist/verbs.json');
const Gerunds = require('english-verbs-gerunds/dist/gerunds.json');

const VerbsData = EnglishVerbs.mergeVerbsData(Irregular, Gerunds);

export function getConjugation(verb, tense: string, person: number) {
      return EnglishVerbs(VerbsData, verb, tense, person <= 2 ? 'S' : 'P')
}
```

es/index.ts:
```ts
export { getConjugation } from 'spanish-verbs';
```

Note that this works only for verb conjugation, and only because Spanish and English are mostly regular.

This break down when trying to scale this up for nouns and adjectives. This is where being able to resolve the used resources comes in handy.

rosaenlg-api:
I understand you would be proposing this module as a way to expose the right functionality? The devil is in the details here, as there is an api for node to generate the resources, then an api for the runtime. It's not clear if a whole new module is necessary here. I could think of a few ways of simply exposing a few new functions in the right modules to achieve the goal.


Other interesting issues:
For solving the multilingual use-case, we'd also have to unify the tenses a bit.  It seems like there should be some "least common denominator" tenses between a few languages. There is no intersection between Spanish tenses and English tenses, it seems.

```ts
export type Tense =
    // english
    'SIMPLE_PAST' | 'PAST' |
    'SIMPLE_PRESENT' | 'PRESENT' |
    'SIMPLE_FUTURE' | 'FUTURE' |
    'PROGRESSIVE_PAST' |
    'PROGRESSIVE_PRESENT' |
    'PROGRESSIVE_FUTURE' |
    'PERFECT_PAST' |
    'PERFECT_PRESENT' |
    'PERFECT_FUTURE' |
    'PERFECT_PROGRESSIVE_PAST' |
    'PERFECT_PROGRESSIVE_PRESENT' |
    'PERFECT_PROGRESSIVE_FUTURE' |
    // spanish
    'INDICATIVE_PRESENT' |
    'INDICATIVE_IMPERFECT' |
    'INDICATIVE_PRETERITE' |
    'INDICATIVE_PERFECT' |
    'INDICATIVE_PLUPERFECT' |
    'INDICATIVE_FUTURE_PERFECT' |
    'INDICATIVE_PRETERITE_PERFECT' |
    'SUBJUNCTIVE_PRESENT' |
    'SUBJUNCTIVE_IMPERFECT_RA' |
    'SUBJUNCTIVE_IMPERFECT_SE' |
    'SUBJUNCTIVE_FUTURE' |
    'SUBJUNCTIVE_PERFECT' |
    'SUBJUNCTIVE_PLUPERFECT' |
    'SUBJUNCTIVE_FUTURE_PERFECT' |
    'CONDITIONAL_PRESENT' |
    'CONDITIONAL_PERFECT';
```
But this is probably an issue for another day.

**Comment by ludans** [comment]
I have started a branch called `rosaenlg-no-pug`, taking your example and putting it in new files, and refactoring some stuff in other packages.
See [test.js in rosaenlg-no-pug package](https://github.com/RosaeNLG/rosaenlg/blob/rosaenlg-no-pug/packages/rosaenlg-no-pug/test/test.js).

Main todo:
- try to have less boilerplate code in `test.js`
- try to get rid of `getPugMixins` (even if for your use case you don't need it)
- check what things would look like when packaging for browser usage - **I would be happy to have your help on that**, as well as other feedback on the branch

Some cleanup to do on my side:
- remove all `spy.getPugMixins().value` (not useful)

On the tense stuff, i.e. mapping language tenses to more universal tenses, why not; let's have something like
Spanish INDICATIVE_PRESENT => UNIVERSAL_PRESENT
English SIMPLE_PRESENT => UNIVERSAL_PRESENT
English PRESENT => UNIVERSAL_PRESENT
Would you like to start such lists in another ticket?

**Comment by ncoder** [comment]
Tres bien.

Any way to remove the need to use a spy here? What happens if
sentenceManager.verb returns the output?
"
/*
| il #[+verb(getAnonMS(), {verb: 'chanter', tense:'FUTUR'} )]
| "#{chanson.nom}"
| de #{chanson.auteur} .
*/
"

I presume here that this is what #{chanson.nom} does as an evaluated
expression... Here i'm not too sure what the #[ ] is used for here.

I think for the tense mappings, yes. that should be a separate issue. I'll
open another ticket.

**Comment by ludans** [comment]
I'm happy that you set high expectations :joy:

Well, in the branch I got completely rid of the `spy` stuff.


As a user it works like that:

- create the `NlgLib` object, indicating merely the language
- optionally indicate where linguistic resources are stored, if working browser side
- code the texts to generate - it can be as simple as generating one verb with `nlgLib.sentenceManager.verb` as well as containing thousands of instructions
- get the final texts using `nlgLib.getFiltered()`

A sample in French:
```javascript
const nlgLib = new NlgLib({ language: 'fr_FR' });
nlgLib.sentenceManager.verb(nlgLib.genderNumberManager.getAnonMS(), { verb: 'devenir', tense: 'FUTUR' });
const rendered = nlgLib.getFiltered();
assert.strictEqual(rendered, 'Deviendra');
```

[noPugMode.js](https://github.com/RosaeNLG/rosaenlg/blob/rosaenlg-no-pug/packages/rosaenlg/test/test-rosaenlg/noPugMode.js) is updated.

Things can be improved in the future:
- test extensively which RosaeNLG features really work in "no pug" mode
- improve the syntax (the code was meant to be generated, not to be written)

I'd be happy to have some help to build an example for server packaging and browser rendering (without using Pug templates).

PS: the `#[ ]` is the way (for Pug) to call a template from within a template (and I don't like this syntax).

**Comment by ludans** [comment]
is available in version 3.0.0

**Comment by ncoder** [comment]
Amazing! I'll try it out as soon as I can.
Then we can talk about packaging.
